### PR TITLE
Add support for downloading external dependencies via buck.

### PR DESCRIPTION
**Pull request by raviagarwal7**
- Kotlin home, groovy home and robolectric dir is now being populated via a buck rule rather than actual jar files.
- SHA256 is computed off the downloaded files via gradle. Support for getting sha256 from remote without downloading will be added next.
- External Dependency api changes to return `targetName` and `targetPath` instead of cache name.

Corresponding buck changes required for this:
https://github.com/facebook/buck/pull/2118
https://github.com/facebook/buck/pull/2119
https://github.com/facebook/buck/pull/2120
https://github.com/facebook/buck/issues/2121

### CODE REVIEW

File: .buckconfig [hunk 1]
```diff
@@ -44,4 +44,14 @@
     app_platform = android-27
     cpu_abis = armv7, x86
 
+[download]
+    in_build = true
+    max_number_of_retries = 3
+
+[maven_repositories]
```
Comment (kageiit) [inline .buckconfig]: This should be inferred from okbuck via the repositories dsl and put in the okbuck generated buckconfig

**Comment by raviagarwal7** [comment]
TODO:
- Reuse `maven_coords` for sources
- Use thread pool for calculating sha256
- Use android model for getting list of mvn repositories

File: buildSrc/src/main/java/com/uber/okbuck/core/manager/GroovyManager.java [hunk 2]
```diff
@@ -22,17 +24,20 @@
...
+  @Nullable private Set<ExternalDependency> dependencies;
 
   public GroovyManager(Project rootProject) {
     this.rootProject = rootProject;
   }
 
   @SuppressWarnings("ResultOfMethodCallIgnored")
   public void setupGroovyHome() {
-    String groovyVersion = GroovySystem.getVersion();
+    groovyVersion = GroovySystem.getVersion();
```
Comment (thalescm) [inline buildSrc/src/main/java/com/uber/okbuck/core/manager/GroovyManager.java]: Does this statical method contains any state? Or is coupled to any lifecycle? If not, groovy version could be not nullable and defined entirely at line 31

File: buildSrc/src/main/java/com/uber/okbuck/composer/java/PrebuiltRuleComposer.java [hunk 3]
```diff
@@ -24,41 +25,30 @@
...
-                      .name(dependency.getCacheName()));
-
-              return rulesBuilder.build();
+                      .sha256(DependencyUtils.sha256(dependency.getRealDependencyFile()));
+              if (dependency.hasSourceFile()) {
+                rule.sourcesMavenCoords(dependency.getSourceMavenCoords())
+                    .sourcesSha256(DependencyUtils.sha256(dependency.getRealSourceFile()));
+              }
+              rule.name(dependency.getTargetName());
+              return ImmutableList.of(rule);
```
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/composer/java/PrebuiltRuleComposer.java]: 👍 will fix

File: buildSrc/src/main/java/com/uber/okbuck/core/manager/DependencyManager.java [hunk 4]
```diff
@@ -216,48 +207,32 @@ private void processDependencies(
...
+                } else {
+                  httpFileDependencies.add(dependency);
+                }
+              });
+
+          ImmutableList.Builder<Rule> rulesBuilder = ImmutableList.builder();
+          rulesBuilder.addAll(PrebuiltRuleComposer.compose(httpFileDependencies.build()));
+          rulesBuilder.addAll(PrebuiltRuleComposer.compose(prebuiltDependencies.build()));
+
+          BuckRuleComposer.composeBuckFile(basePath, rulesBuilder.build());
```
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/manager/DependencyManager.java]: working on adding support for binaries which doesn't have a prebuilt rule. will be replaced with `HttpFileRuleComposer`.

File: buildSrc/src/main/java/com/uber/okbuck/core/manager/GroovyManager.java (hunk 2 above)
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/manager/GroovyManager.java]: seems like it is static in ReleaseInfo.java. https://github.com/groovy/groovy-core/blob/master/src/main/org/codehaus/groovy/util/ReleaseInfo.java#L79

will move it out to a static property.

File: buildSrc/src/main/java/com/uber/okbuck/composer/java/PrebuiltRuleComposer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/composer/java/PrebuiltRuleComposer.java]: done

File: buildSrc/src/main/java/com/uber/okbuck/core/manager/DependencyManager.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/manager/DependencyManager.java]: done

File: buildSrc/src/main/java/com/uber/okbuck/core/manager/GroovyManager.java (hunk 2 above)
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/manager/GroovyManager.java]: done

File: buildSrc/src/main/java/com/uber/okbuck/composer/java/LocalPrebuiltRuleComposer.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+import com.uber.okbuck.core.model.base.RuleType;
+import com.uber.okbuck.template.core.Rule;
+import com.uber.okbuck.template.java.NativePrebuilt;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class LocalPrebuiltRuleComposer extends JvmBuckRuleComposer {
+  /**
+   * @param dependencies External Dependencies whose rule needs to be created
```
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/composer/java/LocalPrebuiltRuleComposer.java]: done

File: buildSrc/src/main/java/com/uber/okbuck/core/dependency/DependencyFactory.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
+package com.uber.okbuck.core.dependency;
+
+import com.google.common.base.Strings;
+import com.uber.okbuck.extension.ExternalDependenciesExtension;
+import com.uber.okbuck.extension.JetifierExtension;
+import java.io.File;
+import javax.annotation.Nullable;
+import org.apache.commons.io.FilenameUtils;
+
+public class DependencyFactory {
```
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/dependency/DependencyFactory.java]: done

File: buildSrc/src/main/java/com/uber/okbuck/core/dependency/DependencyUtils.java [hunk 7]
```diff
@@ -75,6 +74,14 @@ public static boolean isConsumable(File file) {
     return FilenameUtils.isExtension(file.getName(), ALLOWED_EXTENSIONS);
   }
 
+  public static String sha256(File file) {
```
[the remark this answers is not in the document]
Comment (raviagarwal7) [reply buildSrc/src/main/java/com/uber/okbuck/core/dependency/DependencyUtils.java]: done

**Comment by raviagarwal7** [comment]
One thing to note here is that buck doesn't support downloading snapshots.

so for a snapshot version okbuck would generate a non-snapshot maven co-ordinates.
`com.butterknife:butterknife:aar:9.0.0-date-time`, but buck would fail to download this.

renaming it to `com.butterknife:butterknife:aar:9.0.0-SNAPSHOT` works but this would be fragile since the snapshot version can change. 

so, basically with downloading support in buck we won't be able to use snapshots unless we support dynamic snapshots.

**Comment by kageiit** [comment]
I think we can ask users to make snapshots local deps if they need to use it moving forward. i.e okbuck can download them ahead of time in a separate task and we can treat them as local deps. It would be upto the consumers to know the cache issues with using snapshots which we can continue using the messaging we have currently for

**Pull request merged**